shotover/src/transforms/kafka/sink_cluster/scram_over_mtls/connection.rs

```diff
@@ -0,0 +1,159 @@
+        authorize_scram_over_mtls: &Option<AuthorizeScramOverMtls>,
+    ) -> ConnectionState {
+        // The delegation token is recreated after `0.5 * delegation_token_lifetime`
+        // Consider what happens when we match that timing for our connection timeout,
+        // in this timeline going from left to right:
+        //
+        //     create token t1      create token t2
+        // |--------------------|--------------------|
+        // |                    ^ all connections created after this point use token t2 instead of token t1
+        // |                                         |
+        // |             token t1 lifetime           |
+        // |-----------------------------------------|
+        // |                                         ^
+        // |                   after this point, connections still alive that were authed with token t1 will be closed by the broker.
+        // |                                         |
+        // |                                         |
+        // |                                         |
+        // |                                 token t2 lifetime
+        // |                    |-----------------------------------------|
+        // |                    ^ all connections created after this point use token t2
+        // |                                         |
+        // |                                         |
+        // |                                         |
+        // |      connection lifetime using token t1 |
+        // |          |--------------------|         |
+        // This case is fine, the connection exists entirely within the lifetime of token t1.
+        // |                                         |
+        // |                                         |
+        // |                                         |
+        // |                            connection lifetime using token t2
+        // |                                |--------------------|
+        // This case is fine, the connection exists entirely within the lifetime of token t2.
+        // |                                         |
+        // |                                         |
+        // |                                         |
+        // |              connection lifetime using token t?
+        // |                    |--------------------|
+        // This case is a race condition.
+        // We could start with either token t2 or t1.
+        // If we start with t1 we could go past the end of t1's lifetime.
+        // To avoid this issue we reduce the size of the connection lifetime by a further 25%
+        //
+        // At low values of delegation_token_lifetime all of this falls apart since something
+        // like a VM migration could delay shotover execution for many seconds.
+        // However for sufficently large delegation_token_lifetime values (> 1 hour) this should be fine.
+        let timeout = authorize_scram_over_mtls
```

Comment:
A minor comment. Since we do this calculation every time the connection state is checked, would it affect the performance?

Reply:
since the timeout is a constant based on delegation_token_lifetime, would it be worth storing it in authorize_scram_over_mtls as well?

Reply:
I had assumed it would compile down to a single instruction. (or at least very few)
But I looked into it just now and found some interesting results.

On the right we have the ASM generated for the current code.
On the left we have the ASM generated when I combine the `mul_f32` into a single `mul_f32(0.5 * 0.75)`
![image](https://github.com/user-attachments/assets/6342f0f3-62d5-493e-8600-cb479620f5a4)
As you can see reducing the number of `mul_f32` calls greatly reduces the number of instructions.

Inspecting the implementation of mul_f32 I can see that multiplying a Duration is actually fairly complicated, so the results do make some sense.
I'll look into moving the multiply into ScramOverMtlsConnection::new and storing the result.

Reply:
Interesting find!

Reply:
After moving the calculation out of the function entirely:
![image](https://github.com/user-attachments/assets/7e1f7df3-4efb-4a64-91ce-1ec2809edf93)